Added player viewport tracking

penev92 opened the pull request:
This was inspired by multiplayer game casting and the need at times to be able to see what the player sees. I hope this will be helpful to game casters be it in a live multiplayer game or from replays. I was also recently told this will be needed for game saves.

Viewport position and size are sent as regular orders by each player, so they can be monitored live by spectators *and* they go into replay files for later reference.

**_Edit 30.03.2020:_**
Things I am going to need help with:
 - [ ] Figuring out what to do regarding the choppy camera tracking (that is a result of how the orders with the coordinates are sent)
 - [ ] Polishing the UI in TD
 - [ ] Doing the UI in RA (images of what it used to look like can be found below and the corresponding version of the code can be found [in this commit](https://github.com/penev92/OpenRA/commit/9bf71e26219c22e40649be3b44f84edd77144f7e))
 - [ ] Finalize the overall UI layout for this (Also people have mentioned different ideas about the layout entirely, so the overall layout should be finalized sooner rather than later. That should probably also include an option to disable player viewports being drawn on the minimap (another checkbox?))

penev92 committed:
Added trait ViewportTracker and enabled displaying of players' viewports on the minimap by spectators

penev92 committed:
Enabled viewport tracking for the official mods

penev92 committed:
Added tracking of a player's viewport by spectators to enable "though-the-player's-eyes"/"player-view" casting/spectating

obrakmann commented:
First impression is that this is a nice feature. Will be helpful as a learning/teaching tool.

Two suggestions after playing with it for one skirmish and one mission:
1. I think this should be optional, maybe a checkbox in the spectator UI to quickly switch between player-locked and free mode.
2. When it is enabled, viewport modifications by the spectator should be prevented, otherwise it looks unpolished.

penev92 committed:
Added hotkey for observers to lock to a player's camera

penev92 committed:
Made CheckboxWidget's Background configurable

penev92 committed:
Added a several new properties to the scripting API to allow for better control over the player's viewport

penev92 committed:
Made use of the new Camera properties in the scripting API to improve user experience on several missions

penev92 commented:
- Added a checkbox in the observer UI to make locking to a player view optional
 - Added an option to lock the Viewport to disable user interference when following a player's view
 - Added a couple of new properties to the scripting API to take advantage of the "new and improved" Viewport
 - Took advantage of the new scripting API in a couple of missions

This has gotten a bit lengthy, so if people want I'll split it in 2/3 smaller PRs.

dragunoff commented:
Great job @penev92 ! 

However it doesn't work for spectators that used to be in the game but surrendered. The chrome UI is there but it does nothing and no viewport rectangles are shown for other players. I guess this is because `LocalPlayer` is not `null` in that case.

GraionDilach commented:
Cool, scripting API for ingame cinematics!

penev92 commented:
Thanks.
Apparently I've broken the locking in *some* cases (not even all).
I'll fix that and address the comment by @dragunoff , but this is pretty much reviewable in the mean time.

SuperUsername3000 commented:
This feature has been rejected in 2014 : https://github.com/OpenRA/OpenRA/issues/7002. 😐
Nice to finally see it !

pchote commented:
Do you have a timeline for revisiting this @penev92?

reaperrr commented:
Unfortunately there haven't been any updates, so for now I'm closing this too free up the PR queue. Hopefully this will be picked up and finished eventually. Moving to Future milestone.

penev92 committed:
Added a new SetZoom method to Viewport

penev92 committed:
Added locks to Viewport to prevent unwanted manipulations

penev92 committed:
Added a checkbox for observers to enable locking to a player's camera

penev92 committed:
DUMP  Gonna need to figure this out

Added a BackgroundWidget - OBSERVER_CONTROL_BG to combine the widgets under the radar into a single container.

penev92 commented:
Reopening this after a "short" break. Rebasing went horrible and I'll need help with polishing the TD UI and finishing the RA one. D2k and TS turned out nicely, I think. If someone wants to help with the UI, here is the old commit that I believe used to work if this helps: https://github.com/penev92/OpenRA/commit/9bf71e26219c22e40649be3b44f84edd77144f7e

I don't believe I have addressed @dragunoff 's previous comment yet, but since this turned into quite the ordeal that I'm not even quite sure of yet, I wanted to get the ball rolling.

dragunoff commented:
I did some testing with 2 clients and a few bots. I tested with one player being a spectator from the start and with surrendering. Here are my findings:

1. When a player disconnects his viewport rectangle remains but the checkbox no longer locks to it:
![d2k-disconnected-player-viewport](https://user-images.githubusercontent.com/1355810/77639544-9aedad80-6f61-11ea-9bfa-0a66e70aeb1a.png)

1. When a player disconnects while the viewport was locked to their view it becomes permanently locked (the checkbox can be toggled but to no effect):
![d2k-disconnected-locked-viewport](https://user-images.githubusercontent.com/1355810/77639832-0899d980-6f62-11ea-801f-77fa66e8d53b.png)

1. If a player becomes an observer by surrendering then viewport locking has no effect and the viewport rectangles of other players are not displayed (and the player has the option to lock to their own player viewport but again to no effect). This is the same as my comment earlier (https://github.com/OpenRA/OpenRA/pull/16186#issuecomment-462550924):
![d2k-surrendered-own-viewport](https://user-images.githubusercontent.com/1355810/77640141-8c53c600-6f62-11ea-833a-619c4ba63897.png)

1. If a player is defeated their viewport rectangle remains and there is still the option to lock to it. That seems a little meaningless as that player is no longer in the game and we're just seeing their last viewport position:
![d2k-defeated-player-viewport](https://user-images.githubusercontent.com/1355810/77640671-7f83a200-6f63-11ea-82e4-69100eb89a12.png)

1. The checkbox is available for bots even though they don't have a viewport. It should be disabled:
![d2k-bot-viewport](https://user-images.githubusercontent.com/1355810/77640824-b35ec780-6f63-11ea-82be-f92711f84521.png)

1. On the UI side of things - D2k and TS look good, TD should be easy to fix as there's simply rectangles but RA would need some work:
![ra-observer-viewport-ui](https://user-images.githubusercontent.com/1355810/77640701-8ca09100-6f63-11ea-9231-0e190b47bd5b.png)

I also noticed that the movement of the tracked viewport is not smooth and there's a short delay but I guess that's due to networking. Great work overall :+1:

dragunoff commented:
One more thing. I tried running a replay recorded from my test games but I get this crash:
```
OpenRA engine version {DEV_VERSION}
Tiberian Dawn mod version {DEV_VERSION}
on map 86e8f16708089a30144178d42fd9313c2d3c122c (Blank Shellmap by Paul Chote).
Date: 2020-03-26 10:48:11Z
Operating System: Linux (Unix 5.3.0.42)
Runtime Version: Mono 6.8.0.105 (tarball Tue Feb  4 21:20:20 UTC 2020) CLR 4.0.30319.42000
Exception of type `System.IO.EndOfStreamException`: Attempted to read past the end of the stream.
  at OpenRA.StreamExts.ReadBytes (System.IO.Stream s, System.Byte[] buffer, System.Int32 offset, System.Int32 count) [0x00021] in <11bdf3a4e1134a6489977a1916a2fe2d>:0 
  at OpenRA.StreamExts.ReadBytes (System.IO.Stream s, System.Int32 count) [0x0001b] in <11bdf3a4e1134a6489977a1916a2fe2d>:0 
  at OpenRA.Network.ReplayConnection..ctor (System.String replayFilename) [0x00074] in <11bdf3a4e1134a6489977a1916a2fe2d>:0 
  at OpenRA.Game.JoinReplay (System.String replayFile) [0x00000] in <11bdf3a4e1134a6489977a1916a2fe2d>:0 
  at OpenRA.Mods.Common.Widgets.Logic.ReplayBrowserLogic.WatchReplay () [0x0002a] in <0e2d318997e84bf5997f6ce0f7303005>:0 
  at OpenRA.Mods.Common.Widgets.Logic.ReplayBrowserLogic+<>c__DisplayClass14_0.<.ctor>b__3 () [0x00000] in <0e2d318997e84bf5997f6ce0f7303005>:0 
  at OpenRA.Mods.Common.Widgets.ButtonWidget.<.ctor>b__42_5 (OpenRA.MouseInput _) [0x00000] in <0e2d318997e84bf5997f6ce0f7303005>:0 
  at OpenRA.Mods.Common.Widgets.ButtonWidget.HandleMouseInput (OpenRA.MouseInput mi) [0x00076] in <0e2d318997e84bf5997f6ce0f7303005>:0 
  at OpenRA.Widgets.Widget.HandleMouseInputOuter (OpenRA.MouseInput mi) [0x000a0] in <11bdf3a4e1134a6489977a1916a2fe2d>:0 
  at OpenRA.Widgets.Ui.HandleInput (OpenRA.MouseInput mi) [0x0001e] in <11bdf3a4e1134a6489977a1916a2fe2d>:0 
  at OpenRA.DefaultInputHandler+<>c__DisplayClass5_0.<OnMouseInput>b__0 () [0x00000] in <11bdf3a4e1134a6489977a1916a2fe2d>:0 
  at OpenRA.Sync.RunUnsynced[T] (System.Boolean checkSyncHash, OpenRA.World world, System.Func`1[TResult] fn) [0x00023] in <11bdf3a4e1134a6489977a1916a2fe2d>:0 
  at OpenRA.DefaultInputHandler.OnMouseInput (OpenRA.MouseInput input) [0x0000d] in <11bdf3a4e1134a6489977a1916a2fe2d>:0 
  at OpenRA.Platforms.Default.Sdl2Input.PumpInput (OpenRA.Platforms.Default.Sdl2PlatformWindow device, OpenRA.IInputHandler inputHandler, System.Nullable`1[T] lockedMousePosition) [0x00227] in <6bf6f27caf174541b88c272e1ccc694f>:0 
  at OpenRA.Platforms.Default.Sdl2PlatformWindow.PumpInput (OpenRA.IInputHandler inputHandler) [0x00006] in <6bf6f27caf174541b88c272e1ccc694f>:0 
  at OpenRA.Renderer.EndFrame (OpenRA.IInputHandler inputHandler) [0x00099] in <11bdf3a4e1134a6489977a1916a2fe2d>:0 
  at OpenRA.Game.RenderTick () [0x001ba] in <11bdf3a4e1134a6489977a1916a2fe2d>:0 
  at OpenRA.Game.Loop () [0x0015a] in <11bdf3a4e1134a6489977a1916a2fe2d>:0 
  at OpenRA.Game.Run () [0x0003c] in <11bdf3a4e1134a6489977a1916a2fe2d>:0 
  at OpenRA.Game.InitializeAndRun (System.String[] args) [0x00010] in <11bdf3a4e1134a6489977a1916a2fe2d>:0 
  at OpenRA.Program.Main (System.String[] args) [0x00044] in <11bdf3a4e1134a6489977a1916a2fe2d>:0 
```

pchote commented:
IMO the normal sized checkbox / label dominates the sidebar UI, and it may look better if we used a smaller font and introduced new glyphs for smaller checkboxes. Does anybody else have an opinion on this?

If we decide to go ahead with that idea it should be in its own PR (that this one can depend on) to avoid scope creeping this one too much.

penev92 commented:
I guess I'll have to address all the game-ending scenarios that I've been missing.

The live tracking isn't smooth due to networking, yes. That's an issue I raised when I first wrote the code an year and a half ago, but unfortunately the subject gets overshadowed by checkboxes :P
I'm thinking if I should split this into 2 PRs, moving all the UI work into a follow-up PR of the functionality one. Then maybe what @pchote suggested can be a dependency for the second PR, but not the first.

Also your crash doesn't seem to be related to this PR, @dragunoff  :?

PunkPun commented:
> The live tracking isn't smooth due to networking, yes. That's an issue I raised when I first wrote the code an year and a half ago

You can follow the path of sc2 and don't follow the camera directly. If camera moves a distance smaller than a predefined amount then you smoothly move it between those positions. If it is bigger then teleport the camera

penev92 commented:
@dragunoff for reference, this is what the RA UI looked in my original branch:
RA ingame:
![image](https://user-images.githubusercontent.com/7137365/77849347-aea04a80-71d3-11ea-8aec-ac0372dd01e8.png)
RA replay:
![image](https://user-images.githubusercontent.com/7137365/77849371-d4c5ea80-71d3-11ea-904a-7233e6dc1827.png)

Apparently TD looked the same as I got it to work here, but I changed it before reopening the PR. Still can't decide which looks better, but that's just some final polish.
Previous look:
![image](https://user-images.githubusercontent.com/7137365/77849466-72211e80-71d4-11ea-9a1e-3f439e7fe92f.png)
(for comparison, now the replay speed controls (in TD!) don't have a separate background, so no separator line above them).

I'm pretty sure TS and D2k UIs were unaffected by the rebase and look as they did before (*and are "done"*).


I will see what I can do to address the comments regarding player losing/disconnecting/etc. and the movement smoothness, but I've had it with the RA UI so I am going to need help with that one.

abcdefg30 commented:
What does the checkbox do if you are in the `Disable Shroud` view? Wouldn't it make sense to have the checkbox as part of the player dropdown, so that you can select which player to lock onto?

penev92 commented:
Checkbox should be disabled when no particular player is selected. I'm not opposed to moving it, it was just a suggestion, but in that case I'd like to deffer all UI work for a follow-up PR and leave this one with only the main feature.

penev92 committed:
Added checks for player disconnecting to viewport tracking for spectators

penev92 commented:
Started addressing the comments left by @dragunoff.
I fixed 1. and 2. (disconnected players) and 5. (bots). I'll add a fox for 3. (the local player became an observer), but I'm not sure 4. (player becomes an observer) is completely meaningless since the player is still in-game and *can* in fact move their camera around (I am good either way and a fix would be simple, so whatever people want).

Currently I have no solution for the choppy camera tracking. What @Punsho suggested results in the spectator's viewport lagging behind the player's (well it currently also lags behind). I remember back when I wrote the original code there was a big question about Immediate vs not Orders, but I don't remember.
I would like to have this settled before we proceed with any more polishing, so I'm updating the PR description with what I need help with.

dragunoff commented:
@penev92 Point 4. (a defeated player becoming an observer) raises the question of observer viewports in general - should locking to them be allowed? If we allow locking to a defeated player's view then what about locking to other spectators? IMO for the sake of simplicity we should restrict camera locking only to active players' view. 

With regards to the UI - I would agree with @pchote that the checkbox kinda dominates and I won't be opposed to introducing smaller checkboxes but this requires some more thought. I am not a fan of putting mixed content in a dropdown but @abcdefg30's idea might be worth considering as well. In any case I'd suggest we leave the UI details and polish to another PR and just get this one to an acceptable state UI-wise.

pchote commented:
I'm not a fan of merging a feature while leaving major UI questions unresolved - this adds a roadblock that prevents us from doing any releases until the problem is solved.

What if we focused this PR just on showing the rectangles in the minimap, deferring *all* of the locking UI to a single followup?

pchote commented:
Re observer viewports: we should not be exposing these because we want to remove observer orders from the lockstep loop to help with lag.

penev92 commented:
@dragunoff I considered that point last night as well. Figured it could be a nice future improvement to have "directed camera", meaning lock to another spectator and just watch the game (almost like a real e-sports ready game, eh? ;) ). But definitely not here, as that would include a bunch more UI work that is out of scope.

@pchote I think the original idea was just to have rectangles on the map; the locking came in later as a feature that relied on the first one, so can definitely go into its own PR. That would also allow people to focus on the important question of "can this even be made to not lag behind or be choppy so it can be viable", which only touches on how the orders are issued/resolved.

pchote commented:
There is a simple solution for improving the viewport choppyness: clients send an x/y scroll rate in the `ViewportState` order, and the observers use this to forward predict positions until the next tick.

penev92 commented:
I dismissed predictions because that can look very weird when they're wrong though.

pchote commented:
predictions are wrong some of the time, and show a jump back to the actual position. Having the viewport lag / jump between ticks is wrong *all* of the time.

pchote commented:
I implemented a prototype showing how this can work using immediate orders and prediction in https://github.com/pchote/OpenRA/commits/viewport-display. This is much smoother ingame, but currently only works for keyboard-based scrolling.

pchote commented:
Moving to the future milestone to help clean up the PR queue. This is a cool feature, but I don't have the mental bandwidth to take this over myself.
